site/content/blog/multi-env-deployment-ytt-kapp.md

```diff
@@ -0,0 +1,337 @@
+---
+title: "Handle multiple environments with ytt and kapp"
+slug: multi-env-deployment-ytt-kapp
+date: 2022-03-10
+author: Yash Sethiya
+excerpt: "Managing Multi-environments configuration with ytt and kapp"
+image: /img/logo.svg
+tags: ['carvel', 'kapp', 'ytt']
+
+---
+
+One of the most typical challenges when deploying a complex application is the handling of different deployment environments during the software lifecycle.
+
+Commonly, the setup is a trilogy of QA/Staging/Production environments. An application developer needs an easy way to deploy to the different environments and also to understand what version is deployed where.
+
+Unlike [many other tools used for templating](https://carvel.dev/ytt/docs/v0.40.0/ytt-vs-x/), ytt takes a different approach to work with YAML files. Instead of interpreting YAML configuration as plain text, it works with YAML structures such as maps, lists, YAML documents, scalars, etc. By doing so ytt is able to eliminate a lot of problems that plague other tools (character escaping, ambiguity, indentation, etc.). Additionally ytt provides Python-like language (Starlark) that executes in a hermetic environment making it friendly, yet more deterministic compared to just using general purpose languages directly or non-familiar custom templating languages.
+
+
+## How to handle different environment configurations
+
+This problem is usually solved in two ways: templating or patching. [ytt](https://carvel.dev/ytt) supports both approaches. In this section, we’ll see how ytt allows to template YAML configuration, and how it can patch YAML configuration via overlays.
+
+### Data values
+Data values provide a way to inject input data. In ytt, before a Data Value can be used in a template, it must be declared. This is done via [Data Values Schema](https://carvel.dev/ytt/docs/v0.40.0/how-to-write-schema/#overview). We will have a common schema file for all the environments.
+
+```yaml
+#! schema.yaml (#! is used for comment in ytt)
+#@data/values-schema
+---
+replicaCount: 1
+
+image:
+  name: nginx
+  tag: "1.14.2"
+
+nameOverride: ""
+fullnameOverride: ""
```

Comment:
this is not really aligned with general ytt recommendations of what to expose via data values.

Reply:
cc @jtigger

Reply:
I don't know what you mean, @cppforlife.

What would _you_ recommend?

Reply:
@cppforlife made changes as per your feedback.